sherpa/astro/utils/src/_region.cc

```diff
@@ -42,14 +50,32 @@ static PyObject* region_subtract( PyRegion* self, PyObject* args, PyObject *kwar
 static PyObject* region_invert( PyRegion* self, PyObject* args );
 
 
+// The string handling depends on whether we have access to the
+// CIAO data model, in which we case we can use dmRegParse and
+// so support FITS region files, or if we do not have it, in
+// which case we are restricted to ASCII region files via
+// regReadAsciiRegion.
+//
 static regRegion* parse_string( char* str, int fileflag ) {
 
   regRegion *reg = NULL;
+
+#ifdef USE_CXCDM_PARSER
+
+  std::string input(str);
```

Comment:
Since str is a char* - initializing string class variable 'input' with 'str' can segv if the supplied value for str in parse_string() is set to NULL. The code needs to check that str is  non-NULL before this assignement.

Reply:
So, this part is just a slight re-wording of the code that CIAO was using, which didn't include any error checking. Fortunately, in this case, as `parse_string` isn't exported and only used in one place (`pyRegion_New`), where the string argument is created. Thanks to the way that `PyArg_ParseTupleAndKeywords` works when given a `s` format argument we are guaranteed that `objS` isn't `NULL`, so the call to `reg = parse_string( objS, fileflag );` isn't a problem.

I can add some validation though as this is not a part of the code where we care performance or excess memory.

Reply:
This is my proposed change:

```
diff --git a/sherpa/astro/utils/src/_region.cc b/sherpa/astro/utils/src/_region.cc
index 04ce29a3..f15bcf83 100644
--- a/sherpa/astro/utils/src/_region.cc
+++ b/sherpa/astro/utils/src/_region.cc
@@ -56,13 +56,12 @@ static PyObject* region_invert( PyRegion* self, PyObject* args );
 // which case we are restricted to ASCII region files via
 // regReadAsciiRegion.
 //
-static regRegion* parse_string( char* str, int fileflag ) {
+static regRegion* parse_string( std::string input, int fileflag ) {

   regRegion *reg = NULL;

 #ifdef USE_CXCDM_PARSER

-  std::string input(str);
   if( fileflag )
     input = "region(" + input + ")";

@@ -71,9 +70,9 @@ static regRegion* parse_string( char* str, int fileflag ) {
 #else

   if( fileflag ) {
-    reg = regReadAsciiRegion( str, 0 ); // Verbosity set to 0
+    reg = regReadAsciiRegion( (char*)input.c_str(), 0 ); // Verbosity set to 0
   } else {
-    reg = regParse( str );
+    reg = regParse( (char*)input.c_str() );
   }
 #endif

@@ -109,7 +108,12 @@ static PyObject* pyRegion_new(PyTypeObject *type, PyObject *args,
                                     &objS, &fileflag ) )
     return NULL;

-  reg = parse_string( objS, fileflag );
+  // The string is not NULL (otherwise the PyArg_ParseTupleAndKeywords
+  // call would have failed) so we don't need to check here.
+  //
+  std::string name(objS);
+
+  reg = parse_string( name, fileflag );
   if (!reg) {
     const char *fmt;
     if (fileflag) {
```

Is this okay or would you like an explicit check (or have this done some other way)?

Reply:
For reference the description of the `format` argument to `PyArg_ParseTupleXXX` can be found at https://docs.python.org/3.8/c-api/arg.html#strings-and-buffers